### PR TITLE
New 3D-deltaPDF algorithm

### PR DESCRIPTION
**Pull request by rosswhitfield**
This add a new algorithm DeltaPDF3D that will calculate the 3D-deltaPDF from a HKL MDHistoWorkspace

**To test:**
Here are some example MDHisto you can use for testing: `/SNS/users/rwp/corelli/IPTS-16617-PMN/PMN_normdata_48sym_300K.nxs`, `/SNS/users/rwp/corelli/IPTS-18149-Dy2Ti2O7/DTO_normdata_48sym_680mK_H0.8T.nxs`, `/SNS/users/rwp/benzil/benzil_300K_bkg_subtract_sym_All_noCC_flat2.nxs`

Run the DeltaPDF3D algorithm on these (read the docs) and see that something comes out the other side.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by rosswhitfield** [comment]
I just found a problem with the indexing order of `CreateMDHistoWorkspace` being different to how the signalArray is exposed, so I will need to fix that first.

File: Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                logger.information('Using space group: '+space_group)
+                sg=SpaceGroupFactory.createSpaceGroup(space_group)
+            else:
+                check_space_group = False
+            for h in range(int(np.ceil(dimX.getMinimum())), int(np.floor(dimX.getMaximum()))+1):
+                for k in range(int(np.ceil(dimY.getMinimum())), int(np.floor(dimY.getMaximum()))+1):
+                    for l in range(int(np.ceil(dimZ.getMinimum())), int(np.floor(dimZ.getMaximum()))+1):
+                        if not check_space_group or sg.isAllowedReflection([h,k,l]):
+                            if cut_shape == 'cube':
+                                x_min=np.searchsorted(X,h-width[0])
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py]: Is `searchsorted` faster than calculating the index by hand? You know that MDHistoWorkspaces have a regular grid?

File: Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            self.setProperty("IntermediateWorkspace", signalOutWS)
+
+        # Do FFT
+        progress.report("Running FFT")
+        # Replace any remaining nan's or inf's with 0
+        # Otherwise you end up with a lot of nan's
+        signal[np.isnan(signal)]=0
+        signal[np.isinf(signal)]=0
+
+        signal=np.fft.fftshift(np.fft.fftn(np.fft.ifftshift(signal))).real
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py]: Have you looked at pyFFTW? That might be faster

File: Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        createWS_alg.setProperty("Dimensionality", 3)
+        createWS_alg.setProperty("Extents", self._calc_new_extents(inWS))
+        createWS_alg.setProperty("NumberOfBins", number_of_bins)
+        createWS_alg.setProperty("Names", 'x,y,z')
+        createWS_alg.setProperty("Units", 'a,b,c')
+        createWS_alg.execute()
+        outWS = createWS_alg.getProperty("OutputWorkspace").value
+
+        # Copy first experiment info
+        if inWS.getNumExperimentInfo() > 0:
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py]: In principle you can copy all experiment infos.

**Review by AndreiSavici — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                check_space_group = True
+                try:
+                    space_group=SpaceGroupFactory.subscribedSpaceGroupSymbols(int(space_group))[0]
+                except ValueError:
+                    pass
+                logger.information('Using space group: '+space_group)
+                sg=SpaceGroupFactory.createSpaceGroup(space_group)
+            else:
+                check_space_group = False
+            for h in range(int(np.ceil(dimX.getMinimum())), int(np.floor(dimX.getMaximum()))+1):
```
Comment (rosswhitfield) [inline Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py]: No it's not.

File: Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py (hunk 2 above)
Comment (rosswhitfield) [inline Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py]: I will try that out later but I don't really want to add a new dependency to Mantid just for this.

If it is faster I might optionally use pyFFTW over numpy.fft if it's available.

**Review by AndreiSavici — APPROVED** [review]
:shipit:

**Pull request merged**